File: OrangeJuice.Server.Core/Services/XmlRequestValidator.cs
using System;
using System.Xml.Linq;

namespace OrangeJuice.Server.Services
{
	public sealed class XmlRequestValidator : IRequestValidator
	{
		// ReSharper disable once PossibleNullReferenceException
		public bool IsValid(XElement items)
		{
			if (items == null)
				throw new ArgumentNullException("items");

			XNamespace ns = items.Name.Namespace;
			return (bool)items.Element(ns + "Request")
							  .Element(ns + "IsValid");
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Validation/XmlRequestValidator.fs
namespace OrangeJuice.Server.FSharp.Validation

open System.Xml
open System.Xml.Linq
open System.Xml.XPath

open OrangeJuice.Server

type XmlRequestValidator() =
    interface IValidator<XElement> with
        member this.IsValid(item : XElement) : bool =
            if item = null
            then false
            else
                let nm = new XmlNamespaceManager(new NameTable())
                nm.AddNamespace("x", item.Name.Namespace.ToString()) |> ignore

                let toBool e =
                    let o = if e <> null
                                then Some(XElement.op_Explicit e : bool)
                                else None
                    if o.IsSome
                        then o.Value
                        else false

                item.XPathSelectElement("x:Request/x:IsValid", nm) |> toBool
